src/server/tablet_snap.rs

```diff
@@ -101,70 +176,384 @@ fn io_type_from_raft_message(msg: &RaftMessage) -> Result<IoType> {
-) -> Result<u64> {
-    let path = mgr.tablet_gen_path(&key);
-    info!("begin to send snapshot file";"snap_key" => %key);
-    let files = fs::read_dir(&path)?
-        .map(|f| Ok(f?.path()))
-        .filter(|f| f.is_ok() && f.as_ref().unwrap().is_file())
-        .collect::<Result<Vec<_>>>()?;
-    let io_type = io_type_from_raft_message(&msg)?;
-    let _with_io_type = WithIoType::new(io_type);
-    let mut total_sent = msg.compute_size() as u64;
-    let mut chunk = SnapshotChunk::default();
-    chunk.set_message(msg);
-    chunk.set_data(usize::to_le_bytes(SNAP_CHUNK_LEN).to_vec());
-    sender
-        .feed((chunk, WriteFlags::default().buffer_hint(true)))
-        .await?;
-    for path in files {
-        let name = path.file_name().unwrap().to_str().unwrap();
-        let mut buffer = Vec::with_capacity(SNAP_CHUNK_LEN);
-        buffer.push(name.len() as u8);
-        buffer.extend_from_slice(name.as_bytes());
-        let mut f = File::open(&path)?;
-        let mut off = buffer.len();
-        loop {
-            unsafe {
-                buffer.set_len(SNAP_CHUNK_LEN);
+fn protocol_error(exp: &str, act: impl Debug) -> Error {
+    Error::Other(format!("protocol error: expect {exp}, but got {act:?}").into())
+}
+
+async fn cleanup_cache(
+    path: &Path,
+    stream: &mut (impl Stream<Item = Result<TabletSnapshotRequest>> + Unpin),
+    sink: &mut (impl Sink<(TabletSnapshotResponse, WriteFlags), Error = grpcio::Error> + Unpin),
+    limiter: &Limiter,
+) -> Result<(u64, Vec<String>)> {
+    let mut reused = 0;
+    let mut exists = HashMap::default();
+    for entry in fs::read_dir(path)? {
+        let entry = entry?;
+        let ft = entry.file_type()?;
+        if ft.is_dir() {
+            fs::remove_dir_all(entry.path())?;
+            continue;
+        }
+        if ft.is_file() {
+            let os_name = entry.file_name();
+            let name = os_name.to_str().unwrap();
+            if is_sst(name) {
+                // Collect length requires another IO, delay till we are sure
+                // it's probably be reused.
+                exists.insert(name.to_string(), entry.path());
+                continue;
             }
-            // it should break if readed len is zero or the buffer is full.
-            while off < SNAP_CHUNK_LEN {
-                let readed = f.read(&mut buffer[off..])?;
-                if readed == 0 {
-                    unsafe {
-                        buffer.set_len(off);
+        }
+        fs::remove_file(entry.path())?;
+    }
+    let mut missing = vec![];
+    loop {
+        let mut preview = match stream.next().await {
+            Some(Ok(mut req)) if req.has_preview() => req.take_preview(),
+            res => return Err(protocol_error("preview", res)),
+        };
+        let mut buffer = Vec::with_capacity(PREVIEW_CHUNK_LEN);
+        for meta in preview.take_meta().into_vec() {
+            if is_sst(&meta.file_name) && let Some(p) = exists.remove(&meta.file_name) {
```

Comment:
Why is file name enough to vouch for content equalness? Checksum is needed at least.

Reply:
This condition only checks if files with the same name exists, L219~L239 checks for length of file, heading part of the file and trailing part of the file. The heading part contains the actual data of SST, the trailing part contains properties, checksum and other file meta of SST.

Reply:
What's the size of these parts usually? Better comment it here for reference.

Reply:
The size is defined at L72 as constant `PREVIEW_CHUNK_LEN`, which is `1KiB`.